src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixEstimator.java

```diff
@@ -0,0 +1,249 @@
+	// 	NUM_DISTINCT_HYPER_LOG_LOG // HyperLogLog algorithm.
+	// }
+
+	static public int minimumSize = 64000;
+
+	private LibMatrixEstimator() {
+		// Prevent instantiation via private constructor.
+	}
+
+	/**
+	 * Public method to count the number of distinct values inside a matrix. Depending on which EstimatorOperator
+	 * selected it either gets the absolute number or a estimated value.
+	 * 
+	 * TODO: support counting num distinct in rows, or columns axis.
+	 * 
+	 * @param in  the input matrix to count number distinct values in
+	 * @param op  the selected operator to use
+	 * @return the distinct count
+	 */
+	public static int estimateDistinctValues(MatrixBlock in, EstimatorOperator op) {
+		// set output to correct size.
+		
+		// TODO: If the MatrixBlock type is CompressedMatrix, simply read the vaules from the ColGroups.
+
+		if(op.hashType == HashType.ExpHash && op.operatorType == AUType.COUNT_DISTINCT_ESTIMATE_KMV) {
+			throw new DMLException(
+				"Invalid hashing configuration using " + HashType.ExpHash + " and " + AUType.COUNT_DISTINCT_ESTIMATE_KMV);
+		}
+
+		if(op.operatorType == AUType.COUNT_DISTINCT_ESTIMATE_HYPER_LOG_LOG)
+			throw new NotImplementedException("HyperLogLog not implemented");
+
+		// Just use naive implementation if the size is small.
+		if(in.getNumRows() * in.getNumColumns() < minimumSize) {
+			return CountDistinctValuesNaive(in);
+		}
+
+		switch(op.operatorType) {
+			case COUNT_DISTINCT:
+				return CountDistinctValuesNaive(in);
+			case COUNT_DISTINCT_ESTIMATE_KMV:
+				return CountDistinctValuesKVM(in, op);
+			case COUNT_DISTINCT_ESTIMATE_HYPER_LOG_LOG:
+				return CountDistinctHyperLogLog(in);
+			default:
+				throw new DMLException("Invalid or not implemented Estimator Type");
+		}
+	}
+
+	/**
+	 * Naive implementation of counting Distinct values. Benefit Precise, but Uses memory, on the scale of input.
+	 * 
+	 * @param in  the input matrix to count number distinct values in
+	 * @return the distinct count
+	 */
+	private static int CountDistinctValuesNaive(MatrixBlock in) {
+		// Make a hash set to contain all the distinct.
+		// Memory usage scale linear with number of distinct values.
+
+		Set<Double> distinct = new HashSet<Double>();
+
+		for(int c = 0; c < in.getNumColumns(); c++) {
+			for(int r = 0; r < in.getNumRows(); r++) {
+				distinct.add(in.getValue(r, c));
+			}
+		}
+
+		return distinct.size();
+
+		// debugging
+		// LOG.debug("Distinct actual array: " + Arrays.toString(distinct.toArray()));
+		// LOG.debug("Number destinct: " + out);
+
+	}
+
+	/**
+	 * KMV synopsis(for k minimum values) Distinct-Value Estimation
+	 * 
+	 * Kevin S. Beyer, Peter J. Haas, Berthold Reinwald, Yannis Sismanis, Rainer Gemulla: On synopses for distinct‐value
+	 * estimation under multiset operations. SIGMOD 2007
+	 * 
+	 * @param in
+	 * @return the distinct count
+	 */
+	private static int CountDistinctValuesKVM(MatrixBlock in, EstimatorOperator op) {
```

Comment:
Besides the extensions for row/col aggregates, we should also make a TODO for multi-threaded operations, which would require explicit partitioning or a thread-safe priority queue.